packages/xstate-react/README.md

```diff
@@ -70,7 +70,53 @@ A [React hook](https://reactjs.org/hooks) that subscribes to state changes from
 - `state` - Represents the current state of the service as an XState `State` object.
 - `send` - A function that sends events to the running service.
 
-### `useMachine(machine)` with `@xstate/fsm` <Badge text="1.1+"/>
+### `asEffect(action)`
```

Comment:
Right now this won't work at all for invoked/spawned machines. I have no good idea on how to solve this just yet in a way that would not be involving submachines somehow reaching to the "root" of the created machine-tree. And even with something like this I have a problem of figuring out how to perform this action and keeping XState's core not aware of `@xstate/react`'s special needs.

Reply:
I am completely fine with this not working for invoked/spawned machines. Only the immediate machine passed into `useMachine(machine)` should be concerned with actions executed in `useEffect`/`useLayoutEffect`, otherwise we're trying to control absolutely everything that can happen anywhere, which is unrealistic.

Reply:
If this is not intended to work then we should try to implement a warning if somebody tries to do this. 

However, I would be quite surprised that I wouldn't be able to use this for a nested machine in an example like this: https://codesandbox.io/embed/33wr94qv1 (the official TODO example). This limitation is not apparent at all to a user, so we should at least try to surface it is a problem.

Reply:
Actually, now that I think about it, it _can_ work as intended for nested machines. All that `asEffect` and `asLayoutEffect` does is wrap an action in a special function, so if we change `useService` to detect those actions too, we can schedule those in `useEffect` and `useLayoutEffect` all the same.

Reply:
I'll do this in the follow-up PR to refactor `useService` and add `useActor`.

Reply:
Wasnt the overall plan to drop `useService` in favor of the `useActor` which would provide a single way of consuming all actors?

Reply:
Yes, you're right.

Reply:
Then `useActor` couldn't be used to make this work for nested machines - one thing is that it probably shouldn't know anything about what it receives from being subscribed and thus it shouldn't special case received `state.actions`. The other thing is that you can have nested machines without actually using `useService`/`useActor` on them, and you could also call `useService`/`useActor` multiple times on a single service - and we wouldn't like to fire those effects multiple times.

Reply:
Thinking about this more: this is an invalid use-case. Since `useService` is a shared single instance of a started machine, and can be potentially used in many components, we cannot "schedule" the same effect to run at different times for each component; that effect should only fire once. 

If the service is created as `useMachine` in some component, it is _that_ component's responsibility to schedule actions via `asEffect` or `asLayoutEffect`.

What we should do here is add a helper hook for allowing components to _react_ to actions from `useService`; in other words, "I observed that the service did this action, so I will do this action as a reaction to that".

Reply:
Here's what that hook would essentially do:

```js
const inputRef = useRef(null);
const [state, send] = useService(someService); // or useActor, doesn't matter

// hook implementation 
const actionsMap = useMemo(() => ({
  focus: () => { inputRef.current?.focus() }
}, [/* ... */]);

useEffect(() => {
  state.actions.forEach(action => {
    actionsMap[action.type]?(state.context, state.event, {
      state,
      _event: state._event,
      action
    })
  });
}, [state]);
```

Reply:
nit: remember that you can't rely on `state` like that to fire actions, this has to be scheduled through `onTransition` or similar